### ISSUE TITLE
MapView with rounded corners

### ISSUE THREAD

**Issue by AbduazizKayumov** [osmdroid/osmdroid#958]
## Issue Type
[x] Question
[ ] Bug
[x] Improvement
[ ] Build system related
[ ] Performance
[ ] Documentation

## Description and/or steps/code to reproduce the problem
Is there any way that I can achieve `MapView` with rounded corners like `RoundedRectImageView`. I wanted to display the map with 16dp rounded corners (I hate doing like that but it is a design requirement). I tried to get `bitmap` from `MapView` and display it as rounded rectangle drawable. However, the solution does not look intuitive and map coordinates may change anytime (very hard to control). Thank you for any suggestion.

**Comment by monsieurtanuki** [comment]
Quickly said: if you don't need alpha transparent corners but plain gray corners, you may just add the corners on top of the map view's overlays. Something like `CornerOverlay` that would display a "corner" bitmap on each wanted corner of the map view. Some overlays are positioned regardless of the map latitude/longitude, for instance the scale bar.
Hope this helps.

**Comment by MKergall** [comment]
Even simpler, put corner bitmaps suggested by @monsieurtanuki in the XML layer.

**Comment by AbduazizKayumov** [comment]
@monsieurtanuki @MKergall Thank you for suggestions, I would happily display 'gray' bitmaps on each corner of `MapView`, but they should be transparent. `MapView` is inside `RecyclerView.ViewHolder`, and every item is like a chat bubbles with rounded corners. It would be very handy if there was a wrapper `ViewGroup` with rounded corners (similar to setting background to `CardView`).

**Comment by monsieurtanuki** [comment]
@AbduazizKayumov Understood. I'm afraid I cannot help you as I don't know much about this specific subject (`CardView` and so on), though I must say I'd like to know how you will finally manage to do it.

That being said, as an old timer I would make another suggestion: export the bitmap of each wanted map view into a bitmap, then edit the bitmap setting the corner pixels to a fully alpha transparent value (something like 0XFF000000). And display the bitmap as an image.
Of course that doesn't work if you want your map views to be interactive, but in your snapshot example it didn't look so.

**Comment by MKergall** [comment]
Lot of discussions about round corner views here: https://stackoverflow.com/questions/26074784/how-to-make-a-view-in-android-with-rounded-corners

Try either RoundedCornerLayout or CardView?

**Comment by AbduazizKayumov** [comment]
@monsieurtanuki Thank you for suggestion. Actually, it is my 'last hope' plan to achieve rounded corners.
@MKergall I tried both `CardView` and `RoundedCornerLayout` (from the SO answer). `CardView` didn't work as I expected since `MapView` still looks rectangle shaped. `RoundedCornerLayout` draws the rounded corners, but it somehow blocks `MapView` from loading map contents (streets, buildings) and only shows a map with empty rectangles. Also, `RoundedCornerLayout` makes it hard to scroll in case of `RecyclerView`. 
Let me modify `MapView` to achieve rounded corners according to this:
https://stackoverflow.com/questions/13517221/custom-layout-that-rounds-the-corners-of-its-content/32765455
I will share some code snippet if I will have any success on this issue.

**Comment by monsieurtanuki** [comment]
@AbduazizKayumov Any luck?

**Comment by AbduazizKayumov** [comment]
@monsieurtanuki Sorry for not updating the thread for long time. For those who want rounded cornered map, I suggest the following ways:
1. Get drawing cache of `MapView`, convert it to rounded cornered `bitmap`.
2. Modifying `MapView` with this [solution](https://stackoverflow.com/questions/13517221/custom-layout-that-rounds-the-corners-of-its-content/32765455).

The first suggestion is simpler than the second one, get a snapshot from `MapView`, convert it to rounded drawable, may be in this way (in Kotlin):
```
import android.support.v4.graphics.drawable.RoundedBitmapDrawableFactory
// other imports here
object BitmapUtils{
    fun bitmapToRoundedDrawable(): Drawable? {
        val roundedBlur = RoundedBitmapDrawableFactory.create(Pager.instance.resources, this)
        roundedBlur.cornerRadius = SizeUtils.dp2px(16F).toFloat()
        return roundedBlur
    }
}
```
Make sure that this solution adds another computation to your logic, so if you are not dealing with big bitmaps, it should be OK even using it inside `RecyclerView`.

2. I did not test it myself, however, I tried to use `MapView` inside `RecyclerView`. Since the items of `RecyclerView` can be a lot (in my case), the scrolling behavior is not as smooth as I expected (even with `setHasTransientState(true)`). 

Another solution can be using Google's `MapView` with `map:liteMode="true"`. [A lite mode map is a bitmap image of a map at a specified location and zoom level](https://developers.google.com/maps/documentation/android-api/lite). 
I would be very happy to use OSM `MapView` in a lite mode because my first intention was being not dependent to Google Play Services.
This issue may be closed, I see that it is not an issue related to osmdroid.
